File: executors/python-sqlalchemy/app/execute.py
#!/usr/bin/env python3
"""
SQLAlchemy executor for DryORM
Executes user code and tracks queries with line numbers
"""
import os
import sys
import json
import base64
import zlib
import traceback
import io
from contextlib import redirect_stdout
from typing import Any, Dict, List, Optional
import sqlparse
from sqlalchemy import create_engine, event, inspect
from sqlalchemy.engine import Engine
from sqlalchemy.orm import sessionmaker, Session, DeclarativeBase


class LineAwareQueryLogger:
    """Tracks SQL queries with their source line numbers"""

    def __init__(self, user_code_path: str):
        self.queries: List[Dict[str, Any]] = []
        self.user_code_path = user_code_path
        self.ddl_queries: List[str] = []

    def log_query(self, conn, cursor, statement, parameters, context, executemany):
        """Event handler for query execution"""
        import time
        import linecache

        # Get the stack trace to find where the query originated
        stack = traceback.extract_stack()

        # Find the frame from user code
        line_number = None
        source_context = None

        for frame in reversed(stack):
            if self.user_code_path in frame.filename:
                line_number = frame.lineno
                source_context = linecache.getline(frame.filename, frame.lineno).strip()
                break

        # Format the SQL
        formatted_sql = sqlparse.format(
            str(statement),
            reindent=True,
            keyword_case='upper'
        )

        query_info = {
            'sql': formatted_sql,
            'template': str(statement),
            'time': '0.000',  # SQLAlchemy doesn't provide query time in this event
            'line_number': line_number,
            'source_context': source_context
        }

        self.queries.append(query_info)

        # Track DDL statements for ERD generation
        statement_upper = str(statement).upper().strip()
        if any(statement_upper.startswith(ddl) for ddl in ['CREATE TABLE', 'CREATE INDEX', 'ALTER TABLE']):
            self.ddl_queries.append(formatted_sql)


def generate_erd(engine: Engine, base_class: Optional[type] = None) -> str:
    """Generate Mermaid ERD diagram from SQLAlchemy models"""
    from mermaid_generator import generate_mermaid_erd

    try:
        mermaid = generate_mermaid_erd(engine, base_class)
        # Compress and encode like Django executor
        compressed = zlib.compress(mermaid.encode('utf-8'))
        encoded = base64.b64encode(compressed).decode('utf-8')
        return encoded
    except Exception as e:
        # Return empty string on error
        return ""


def get_database_url() -> str:
    """Build database URL from environment variables"""
    db_type = os.environ.get('DB_TYPE', 'sqlite')

    if db_type == 'sqlite':
        return 'sqlite:///db.sqlite3'

    db_host = os.environ.get('SERVICE_DB_HOST', 'localhost')
    db_port = os.environ.get('SERVICE_DB_PORT', '5432')
    db_user = os.environ.get('DB_USER', 'user')
    db_password = os.environ.get('DB_PASSWORD', 'password')
    db_name = os.environ.get('DB_NAME', 'test')

    if db_type == 'postgres':
        return f'postgresql://{db_user}:{db_password}@{db_host}:{db_port}/{db_name}'
    elif db_type == 'mariadb':
        return f'mysql+pymysql://{db_user}:{db_password}@{db_host}:{db_port}/{db_name}'
    else:
        raise ValueError(f"Unsupported database type: {db_type}")


def main():
    """Main execution function"""
    user_code_path = '/app/app/models.py'

    # Create database connection
    database_url = get_database_url()
    engine = create_engine(database_url, echo=False)

    # Set up query logging
    logger = LineAwareQueryLogger(user_code_path)
    event.listen(engine, "before_cursor_execute", logger.log_query)

    # Load user code
    try:
        with open(user_code_path, 'r') as f:
            user_code = f.read()
    except FileNotFoundError:
        print(json.dumps({
            'error': 'models.py not found',
            'output': '',
            'queries': [],
            'erd': '',
            'returned': None
        }))
        sys.exit(1)

    # Set up execution environment
    output_buffer = io.StringIO()
    result = {
        'output': '',
        'queries': [],
        'erd': '',
        'returned': None
    }

    try:
        # Execute user code
        user_globals = {
            '__name__': '__main__',
            '__file__': user_code_path,
            'engine': engine,
            'Session': sessionmaker(bind=engine),
            'DeclarativeBase': DeclarativeBase,
        }

        # Import commonly used SQLAlchemy components for convenience
        from sqlalchemy import (
            Column, Integer, String, Text, Boolean, Float, DateTime, Date, Time,
            ForeignKey, Table, MetaData, select, insert, update, delete,
            and_, or_, not_, desc, asc, func
        )
        from sqlalchemy.orm import relationship, backref, Mapped, mapped_column

        user_globals.update({
            'Column': Column,
            'Integer': Integer,
            'String': String,
            'Text': Text,
            'Boolean': Boolean,
            'Float': Float,
            'DateTime': DateTime,
            'Date': Date,
            'Time': Time,
            'ForeignKey': ForeignKey,
            'Table': Table,
            'MetaData': MetaData,
            'relationship': relationship,
            'backref': backref,
            'Mapped': Mapped,
            'mapped_column': mapped_column,
            'select': select,
            'insert': insert,
            'update': update,
            'delete': delete,
            'and_': and_,
            'or_': or_,
            'not_': not_,
            'desc': desc,
            'asc': asc,
            'func': func,
        })

        # Compile and execute user code
        compiled_code = compile(user_code, user_code_path, 'exec')

        with redirect_stdout(output_buffer):
            exec(compiled_code, user_globals)

            # Look for Base class (DeclarativeBase subclass)
            base_class = None
            for name, obj in user_globals.items():
                if (isinstance(obj, type) and
                    issubclass(obj, DeclarativeBase) and
                    obj is not DeclarativeBase):
                    base_class = obj
                    break

            # Create tables if Base class found
            if base_class:
                base_class.metadata.create_all(engine)

            # Run the user's run() function if it exists
            if 'run' in user_globals and callable(user_globals['run']):
                return_value = user_globals['run']()
                result['returned'] = str(return_value) if return_value is not None else None

        result['output'] = output_buffer.getvalue()
        result['queries'] = logger.queries

        # Generate ERD
        result['erd'] = generate_erd(engine, base_class)

    except Exception as e:
        result['output'] = output_buffer.getvalue()
        result['queries'] = logger.queries
        result['error'] = f"{type(e).__name__}: {str(e)}"
        result['traceback'] = traceback.format_exc()

    finally:
        engine.dispose()

    # Output result as JSON
    print(json.dumps(result))


if __name__ == '__main__':
    main()

File: executors/nodejs-prisma/app/execute.js
#!/usr/bin/env node
/**
 * Prisma executor for DryORM
 * Executes user code and tracks queries with line numbers
 */

import { PrismaClient } from '@prisma/client';
import { readFileSync, writeFileSync } from 'fs';
import { createRequire } from 'module';
import * as path from 'path';
import { fileURLToPath } from 'url';
import { exec } from 'child_process';
import { promisify } from 'util';
import zlib from 'zlib';

const execAsync = promisify(exec);
const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

// Line-aware query logger
class LineAwareQueryLogger {
  constructor() {
    this.queries = [];
  }

  log(query) {
    // Get stack trace to find the source line
    const stack = new Error().stack.split('\n');
    let lineNumber = null;
    let sourceContext = null;

    // Find the frame from user code
    for (const frame of stack) {
      if (frame.includes('/app/user-code.js')) {
        const match = frame.match(/:(\d+):\d+/);
        if (match) {
          lineNumber = parseInt(match[1], 10);
          // Try to get the source line
          try {
            const userCode = readFileSync('/app/user-code.js', 'utf-8');
            const lines = userCode.split('\n');
            if (lineNumber > 0 && lineNumber <= lines.length) {
              sourceContext = lines[lineNumber - 1].trim();
            }
          } catch (e) {
            // Ignore errors
          }
        }
        break;
      }
    }

    this.queries.push({
      sql: query.query,
      template: query.query,
      time: (query.duration || 0).toFixed(3),
      line_number: lineNumber,
      source_context: sourceContext
    });
  }
}

// Database URL builder
function getDatabaseUrl() {
  const dbType = process.env.DB_TYPE || 'sqlite';

  if (dbType === 'sqlite') {
    return 'file:./db.sqlite3';
  }

  const dbHost = process.env.SERVICE_DB_HOST || 'localhost';
  const dbPort = process.env.SERVICE_DB_PORT || '5432';
  const dbUser = process.env.DB_USER || 'user';
  const dbPassword = process.env.DB_PASSWORD || 'password';
  const dbName = process.env.DB_NAME || 'test';

  if (dbType === 'postgres') {
    return `postgresql://${dbUser}:${dbPassword}@${dbHost}:${dbPort}/${dbName}`;
  } else if (dbType === 'mariadb') {
    return `mysql://${dbUser}:${dbPassword}@${dbHost}:${dbPort}/${dbName}`;
  } else {
    throw new Error(`Unsupported database type: ${dbType}`);
  }
}

// Generate ERD from schema
async function generateERD() {
  try {
    // Read schema file
    const schema = readFileSync('/app/schema.prisma', 'utf-8');

    // Parse schema and generate Mermaid ERD
    const mermaid = generateMermaidFromSchema(schema);

    // Compress and encode
    const compressed = zlib.deflateSync(Buffer.from(mermaid, 'utf-8'));
    return compressed.toString('base64');
  } catch (e) {
    return '';
  }
}

// Safe JSON stringifier that handles circular references and BigInt
function safeStringify(obj) {
  const seen = new WeakSet();
  return JSON.stringify(obj, (key, value) => {
    // Handle BigInt
    if (typeof value === 'bigint') {
      return value.toString();
    }
    // Handle circular references
    if (typeof value === 'object' && value !== null) {
      if (seen.has(value)) {
        return '[Circular]';
      }
      seen.add(value);
    }
    return value;
  });
}

// Simple Mermaid ERD generator from Prisma schema
function generateMermaidFromSchema(schema) {
  const lines = schema.split('\n');
  let mermaid = 'erDiagram\n';
  let currentModel = null;
  const models = [];
  const relationships = [];

  for (const line of lines) {
    const trimmed = line.trim();

    // Model declaration
    if (trimmed.startsWith('model ')) {
      currentModel = trimmed.match(/model\s+(\w+)/)?.[1];
      if (currentModel) {
        models.push({
          name: currentModel,
          fields: []
        });
      }
    } else if (currentModel && trimmed && !trimmed.startsWith('//') && !trimmed.startsWith('}')) {
      // Field declaration
      const fieldMatch = trimmed.match(/(\w+)\s+(\w+)(\[\])?(\?)?/);
      if (fieldMatch) {
        const [, fieldName, fieldType, isArray, isOptional] = fieldMatch;
        const model = models[models.length - 1];

        // Check if it's a relation field
        if (models.some(m => m.name === fieldType)) {
          const relationType = isArray ? '||--o{' : (isOptional ? '||--o|' : '||--||');
          relationships.push(`${currentModel} ${relationType} ${fieldType} : ""`);
        } else {
          // Regular field
          const typeStr = fieldType + (isArray ? '[]' : '') + (isOptional ? '?' : '');
          model.fields.push({ name: fieldName, type: typeStr });
        }
      }
    }
  }

  // Generate model definitions
  for (const model of models) {
    mermaid += `  ${model.name} {\n`;
    for (const field of model.fields) {
      mermaid += `    ${field.type} ${field.name}\n`;
    }
    mermaid += `  }\n`;
  }

  // Add relationships
  for (const rel of relationships) {
    mermaid += `  ${rel}\n`;
  }

  return mermaid;
}

// Main execution function
async function main() {
  const result = {
    output: '',
    queries: [],
    erd: '',
    returned: null
  };

  // Declare these at function level so they're accessible in catch blocks
  const originalLog = console.log;
  const logs = [];
  const logger = new LineAwareQueryLogger();
  let prisma = null;

  try {
    // Create Prisma client with query logging
    prisma = new PrismaClient({
      log: [
        {
          emit: 'event',
          level: 'query',
        },
      ],
    });

    // Set up query event listener
    prisma.$on('query', (e) => {
      logger.log(e);
    });

    // Run migrations (redirect output to stderr to avoid polluting JSON output)
    try {
      await execAsync('npx prisma migrate dev --name init --skip-generate >&2', {
        cwd: '/app',
        env: { ...process.env, DATABASE_URL: getDatabaseUrl() }
      });
    } catch (migrationError) {
      // Try db push as fallback
      try {
        await execAsync('npx prisma db push --skip-generate >&2', {
          cwd: '/app',
          env: { ...process.env, DATABASE_URL: getDatabaseUrl() }
        });
      } catch (pushError) {
        // Ignore if already exists
      }
    }

    // Load and execute user code
    const userCodePath = '/app/user-code.js';

    // Capture stdout (but don't write to actual stdout to avoid polluting JSON)
    console.log = (...args) => {
      logs.push(args.map(a => String(a)).join(' '));
      // Don't call originalLog - we only want to capture, not output
    };

    try {
      // Import user code as ES module
      const userModule = await import(userCodePath + '?t=' + Date.now());

      // Call the run function if it exists
      if (typeof userModule.run === 'function') {
        const returnValue = await userModule.run(prisma);
        if (returnValue !== undefined && returnValue !== null) {
          // Handle arrays and objects properly
          if (typeof returnValue === 'object') {
            result.returned = returnValue;
          } else {
            result.returned = String(returnValue);
          }
        }
      }
    } finally {
      console.log = originalLog;
    }

    result.output = logs.join('\n');
    result.queries = logger.queries;

    // Generate ERD
    result.erd = await generateERD();

    // Clean up
    if (prisma) {
      await prisma.$disconnect();
    }

  } catch (error) {
    // Restore console.log in case of error
    console.log = originalLog;

    result.output = logs.join('\n');
    result.queries = logger.queries;
    result.error = `${error.name}: ${error.message}`;
    result.traceback = error.stack;

    // Clean up prisma connection on error
    if (prisma) {
      try {
        await prisma.$disconnect();
      } catch (disconnectError) {
        // Ignore disconnect errors
      }
    }
  }

  // Restore console.log before final output
  console.log = originalLog;

  // Write result to file instead of stdout to avoid pollution
  try {
    const jsonOutput = safeStringify(result);
    const resultFile = '/tmp/result.json';

    writeFileSync(resultFile, jsonOutput);

    // Also write to stdout for backward compatibility (can be removed later)
    process.stdout.write(jsonOutput + '\n');
  } catch (stringifyError) {
    // If even safe stringify fails, output a basic error
    const fallbackResult = {
      output: result.output || '',
      queries: [],
      erd: '',
      returned: null,
      error: `JSON Serialization Error: ${stringifyError.message}`,
      traceback: stringifyError.stack
    };
    const resultFile = '/tmp/result.json';
    writeFileSync(resultFile, JSON.stringify(fallbackResult));
    process.stdout.write(JSON.stringify(fallbackResult) + '\n');
  }
}

main().catch((error) => {
  // Fatal error - output basic error result
  const errorResult = {
    output: '',
    queries: [],
    erd: '',
    returned: null,
    error: `Fatal Error: ${error.message}`,
    traceback: error.stack
  };
  try {
    const jsonOutput = safeStringify(errorResult);
    const resultFile = '/tmp/result.json';
    writeFileSync(resultFile, jsonOutput);
    process.stdout.write(jsonOutput + '\n');
  } catch (e) {
    const fallbackOutput = JSON.stringify({
      output: '',
      queries: [],
      erd: '',
      returned: null,
      error: 'Critical serialization error',
      traceback: ''
    });
    const resultFile = '/tmp/result.json';
    writeFileSync(resultFile, fallbackOutput);
    process.stdout.write(fallbackOutput + '\n');
  }
  process.exit(1);
});
